pyspark_ai/python_executor.py

```
@@ -0,0 +1,91 @@
+        # assert llm is an instance of BaseChatModel
+        assert isinstance(
+            self.llm, BaseChatModel
+        ), "The llm is not an instance of BaseChatModel"
+        prompt_str = canonize_string(self.prompt.format_prompt(**kwargs).to_string())
+        use_cache = tags != SKIP_CACHE_TAGS
+        if self.cache is not None:
+            cached_result = self.cache.lookup(prompt_str) if use_cache else None
+            if cached_result is not None:
+                self._execute_code(self.df, cached_result)
+                return cached_result
+        messages = [HumanMessage(content=prompt_str)]
+        response = self._generate_python_with_retries(
+            self.df, self.llm, messages, self.max_retries
+        )
+        if use_cache and self.cache is not None:
+            self.cache.update(prompt_str, response)
+        return response
+
+    @staticmethod
+    def _execute_code(df: DataFrame, code: str):
+        import pandas as pd  # noqa: F401
+
+        exec(compile(code, "plot_df-CodeGen", "exec"))
+
+    def _generate_python_with_retries(
+        self,
+        df: DataFrame,
+        chat_model: BaseChatModel,
+        messages: List[BaseMessage],
+        retries: int = 3,
+    ) -> str:
+        response = chat_model.predict_messages(messages)
+        if self.logger is not None:
+            self.logger.info(response.content)
+        code = "\n".join(AIUtils.extract_code_blocks(response.content))
+        try:
+            self._execute_code(df, code)
+            return code
+        except Exception as e:
+            if self.logger is not None:
+                self.logger.warning("Getting the following error: \n" + str(e))
+            if retries <= 0:
+                # if we have no more retries, raise the exception
+                self.logger.info(
+                    "No more retries left, please modify the instruction or modify the generated code"
+                )
+                return ""
```

Do we want to raise an exception here instead?

---

I thought about it. The error stack trace will make the logs hard to read, especially with retries.